sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala

```diff
@@ -272,6 +275,83 @@ class ParquetFileFormat
     true
   }
 
+  override def getSplits(fileIndex: FileIndex,
+                         fileStatus: FileStatus,
+                         filters: Seq[Filter],
+                         schema: StructType,
+                         hadoopConf: Configuration): Seq[FileSplit] = {
+    if (filters.isEmpty) {
+      // Return immediately to save FileSystem overhead
+      super.getSplits(fileIndex, fileStatus, filters, schema, hadoopConf)
+    } else {
+      val filePath = fileStatus.getPath
+      val rootOption: Option[Path] = fileIndex.rootPaths
+        .find(root => filePath.toString.startsWith(root.toString))
+      val metadataOption = rootOption.flatMap(getMetadataForPath(filePath, _, hadoopConf))
+      // If the metadata exists, filter the splits.
+      // Otherwise, fall back to the default implementation.
+      metadataOption
+        .map(filterToSplits(fileStatus, _, rootOption.get, filters, schema, hadoopConf))
+        .getOrElse(super.getSplits(fileIndex, fileStatus, filters, schema, hadoopConf))
+    }
+  }
+
+  private def filterToSplits(fileStatus: FileStatus,
+                             metadata: ParquetMetadata,
+                             metadataRoot: Path,
+                             filters: Seq[Filter],
+                             schema: StructType,
+                             hadoopConf: Configuration): Seq[FileSplit] = {
+    val metadataBlocks = metadata.getBlocks
+    val parquetSchema = metadata.getFileMetaData.getSchema
+    val filter = FilterCompat.get(filters
+          .flatMap(ParquetFilters.createFilter(schema, _))
+          .reduce(FilterApi.and))
+    val filteredMetadata =
+      RowGroupFilter.filterRowGroups(filter, metadataBlocks, parquetSchema).asScala
+    filteredMetadata.flatMap(bmd => {
```

Comment:
According to https://github.com/databricks/scala-style-guide#anonymous-methods, 

```scala
filteredMetadata.flatMap { bmd =>
  ...
```

seems encouraged.